### PR TITLE
feat(core): Add configurable expiry to X.509 certificates

### PR DESCRIPTION
**Pull request by ddneilson**
Implements: https://github.com/aws/aws-rfdk/issues/241

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by horsmand — DISMISSED** [review]

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts [hunk 1]
```diff
@@ -88,6 +88,13 @@ export interface X509CertificatePemProps {
    * @default: None. The generated certificate will be self-signed
    */
   readonly signingCertificate?: X509CertificatePem;
+
+  /**
+   * The number of days that the generated certificate will be valid for.
+   *
+   * @default 1095 days (3 years)
+   */
+  readonly validFor?: number;
```
Comment (jusiskin) [inline packages/aws-rfdk/lib/core/lib/x509-certificate.ts:98]: How about using a `cdk.Duration` here? It allows users to be more expressive. Same feedback for `.fromGenerated`.

File: packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts [hunk 2]
```diff
@@ -76,6 +76,32 @@ test('generate self-signed', async () => {
...
+  await writeAsciiFile(crtFileName, certificate.cert);
+  await writeAsciiFile(keyFileName, certificate.key);
+  const certOut = await exec(`openssl x509 -noout -text -in ${crtFileName}`);
+  const certVerification: string = certOut.stdout;
+  const expiryDate = new Date();
+  expiryDate.setDate(expiryDate.getDate() + 5*365);
+
+  // THEN
+  // ex: Not After : May 22 22:13:24 2023 GMT
+  expect(certVerification).toMatch(new RegExp(`Not After.*${expiryDate.getFullYear()} GMT`));
```
Comment (jusiskin) [inline packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts:106]: This only checks the year. Do we care about having more strict validation?

File: packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts [hunk 3]
```diff
@@ -92,7 +118,7 @@ test('generate signed certificate', async () => {
   const passphrase: string = 'test_passphrase';
 
   // WHEN
-  const certificate = await Certificate.fromGenerated(certName, passphrase, ca);
+  const certificate = await Certificate.fromGenerated(certName, passphrase, undefined, ca);
```
Comment (jusiskin) [inline packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts:125]: It appears `undefined` means the default expiry of three years. Should we assert that?

File: packages/aws-rfdk/lib/lambdas/nodejs/x509-certificate/types.ts [hunk 4]
```diff
@@ -165,6 +171,10 @@ export function implementsIX509CertificateGenerate(value: any): boolean {
   if (!implementsIX509ResourceProperties(value)) { return false; }
   if (!implementsDistinguishedNameProps(value.DistinguishedName)) { return false; }
   if (value.SigningCertificate && !implementsISecretCertificate(value.SigningCertificate)) { return false; }
+  if (value.CertificateValidFor) {
+    if (typeof(value.CertificateValidFor) !== 'string') { return false; }
+    if (value.CertificateValidFor === '' || isNaN(Number(value.CertificateValidFor))) { return false; }
```
Comment (jusiskin) [inline packages/aws-rfdk/lib/lambdas/nodejs/x509-certificate/types.ts:176]: Should we add a validation here to ensure we have a positive number?

**Review by jusiskin — CHANGES_REQUESTED** [review]
Looks good. Just a few suggestions

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts (hunk 1 above)
Comment (ddneilson) [reply packages/aws-rfdk/lib/core/lib/x509-certificate.ts:98]: I had the same thought, but `Duration`'s units are at most days (ex: no 'years', 'months', etc). So, it doesn't really seem to add much other than extra syntax for the customer and testing on our side.

File: packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts (hunk 2 above)
Comment (ddneilson) [reply packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts:106]: It's a bit of transparent testing; intent is just to make sure that the days value is getting all the way through to openssl. The default is 3 years, so as long as the year changes by more than 3 then we know that the value was passed through to the openssl command.

I'm also trying to avoid leap-year rounding issues by going for a more coarse-grained check.

File: packages/aws-rfdk/lib/lambdas/nodejs/x509-certificate/types.ts (hunk 4 above)
Comment (ddneilson) [reply packages/aws-rfdk/lib/lambdas/nodejs/x509-certificate/types.ts:176]: Not here, but it makes some sense to assert that in the construct itself. Fail early...

File: packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts (hunk 3 above)
Comment (ddneilson) [reply packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts:125]: No harm in it, sure.

File: packages/aws-rfdk/lib/core/lib/x509-certificate.ts (hunk 1 above)
Comment (jusiskin) [reply packages/aws-rfdk/lib/core/lib/x509-certificate.ts:98]: Ah that makes sense. Resolved

File: packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts (hunk 2 above)
Comment (jusiskin) [reply packages/aws-rfdk/lib/lambdas/nodejs/lib/x509-certs/test/certificate.test.ts:106]: This makes sense. Although we may still encounter leap year rounding issues in the form of false-negatives if we execute these tests on a date near December 31/January 1st. This is a minor edge-case though - so let's just keep it as-is

File: packages/aws-rfdk/lib/lambdas/nodejs/x509-certificate/types.ts (hunk 4 above)
Comment (jusiskin) [reply packages/aws-rfdk/lib/lambdas/nodejs/x509-certificate/types.ts:176]: I'm okay with this approach, although it does miss the case where the customer passes in a CDK token. Sounds like that should be avoided since, as this PR points out, it could lead to certificates being regenerated and breaking any manually configured clients.

On that basis, do you think we should also validate that the expiration is not a token?

Comment (ddneilson) [reply packages/aws-rfdk/lib/lambdas/nodejs/x509-certificate/types.ts:176]: Token should be allowable. That'll get expanded into its value by CloudFormation.

Though, I suppose that means that my validation check isn't quite right. :-/

Comment (ddneilson) [reply packages/aws-rfdk/lib/lambdas/nodejs/x509-certificate/types.ts:176]: Updated in the latest revision. I'm leaving the negative-value test out of the type validator. Those validators do not give useful diagnostic output; they're mostly to make sure that we've created the CustomResource properly.

openssl will die with an error like `req: Non-positive number "-1" for -days` if a negative value is passed through.

Comment (jusiskin) [reply packages/aws-rfdk/lib/lambdas/nodejs/x509-certificate/types.ts:176]: Great - thanks for looking into this.

**Review by jusiskin — APPROVED** [review]

**Review by horsmand — APPROVED** [review]
